spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,223 @@
+  protected List<CombinedScanTask> tasks() {
+    Preconditions.checkState(startOffset != null,
+        "Start offset is null when planning tasks, this probably because there's no snapshot in the table currently");
+    Preconditions.checkState(endOffset != null, "End offset is null when planning tasks");
+
+    StreamingOffset start = (StreamingOffset) startOffset;
+    StreamingOffset end = (StreamingOffset) endOffset;
+
+    List<SnapshotTasks> snapshotTasksList = getChanges(start.snapshotId(), start.index(), start.isStartingSnapshotId());
+    List<CombinedScanTask> runningTasks = Lists.newArrayList();
+    for (SnapshotTasks s : snapshotTasksList) {
+      if (s.snapshotId() == end.snapshotId()) {
+        runningTasks.addAll(s.tasks().subList(0, end.index() + 1));
+        break;
+      } else {
+        runningTasks.addAll(s.tasks());
+      }
+    }
+
+    return runningTasks;
+  }
+
+  private StreamingOffset calculateInitialOffset() {
+    Snapshot current = table.currentSnapshot();
+    List<Long> snapshotIds = SnapshotUtil.ancestorIds(current, table::snapshot);
+    if (snapshotIds.isEmpty()) {
+      return null;
+    } else {
+      long initSnapshotId = snapshotIds.get(snapshotIds.size() - 1);
+      return new StreamingOffset(initSnapshotId, -1, true /* is_starting */);
+    }
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    List<SnapshotTasks> snapshotTasksList = getChangesWithRateLimit(start.snapshotId(),
+        start.index(), start.isStartingSnapshotId());
+    SnapshotTasks last = snapshotTasksList.isEmpty() ? null : snapshotTasksList.get(snapshotTasksList.size() - 1);
+
+    return last == null ? start : new StreamingOffset(last.snapshotId, last.size() - 1,
+        last.snapshotId() == start.snapshotId() /* is_starting */);
+  }
+
+  @VisibleForTesting
+  List<SnapshotTasks> getChanges(long snapshotId, int index, boolean isStarting) {
+    List<SnapshotTasks> snapshotTasksList = Lists.newArrayList();
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+
+    if (isStarting) {
+      CloseableIterable<CombinedScanTask> iter = buildTableScan().useSnapshot(snapshotId).planTasks();
+      SnapshotTasks snapshotTasks = new SnapshotTasks(snapshotId, Lists.newArrayList(iter));
+      snapshotTasksList.add(snapshotTasks);
+    }
+
+    Long startSnapshotId = isStarting ? Long.valueOf(snapshotId) : table.snapshot(snapshotId).parentId();
+    Preconditions.checkState(startSnapshotId != null, "Start snapshot id should be existed");
+    List<Long> snapshotIds = SnapshotUtil.snapshotIdsBetween(table, startSnapshotId, currentSnapshotId);
+    snapshotIds.add(startSnapshotId);
+
+    for (int i = snapshotIds.size() - 1; i > 0; i--) {
+      long fromSnapshotId = snapshotIds.get(i);
+      long toSnapshotId = snapshotIds.get(i - 1);
+      CloseableIterable<CombinedScanTask> iter = buildTableScan()
+          .appendsBetween(fromSnapshotId, toSnapshotId)
+          .planTasks();
```

Comment:
Internally, `planTasks` calls `planFiles`, which uses a worker thread-pool to scan manifests and [add matching files to a concurrent queue](https://github.com/apache/incubator-iceberg/blob/master/core/src/main/java/org/apache/iceberg/util/ParallelIterable.java#L65-L67). When using parallel planning, the order of files is non-deterministic and so are the tasks that end up getting created because the order influences how files are combined into tasks.

It looks like this currently relies on deterministic behavior in `planTasks`. That's because `getChanges` is called more than once -- both while calculating the end offset and planning tasks for a batch. Even if we were to add a cache to avoid the cost of scan planning more than once, this would still be a problem because tasks could be different when recovering from failure if an index is in the middle of a task list for a snapshot.

I think we need to guarantee deterministic behavior for tasks. I think there are two options:
1. Disable parallel scan planning by adding an option to `TableScan`
2. Call `planFiles` instead of `planTasks`, but sort the files to produce a deterministic order. Then, split and combine files into tasks.

I prefer option 2 since it still scans table metadata in parallel. We would just need to move the splitting and combining code from `BaseTableScan` into utility methods.

That should address correctness, but I think we also want to add a cache for the files or splits in each snapshot. We don't want to scan snapshot metadata files more than once if we can avoid it because it is expensive.

Reply:
Thanks for pointing out the key issue. This part is also the most unsure part when I achieve this.  I will think a bit on this and change the current way.

Reply:
> , but I think we also want to add a cache for the files or splits in each snapshot

@rdblue Can I separate this thing into another PR? I was thinking that this is not super related to streaming reader here.

Reply:
Yes, I agree it should be in a separate PR. No need to over-complicate this one.